File: apps/proxy/internal/proxy/attempts_test.go
package proxy

import (
	"context"
	"io"
	"net/http"
	"net/http/httptest"
	"strconv"
	"strings"
	"testing"
	"time"

	"github.com/opendum/opendum/apps/proxy/internal/auth"
	appdb "github.com/opendum/opendum/apps/proxy/internal/db"
)

func TestExecuteAccountRotationContinuesPastFiveFailures(t *testing.T) {
	runner := &testRotationRunner{accounts: []appdb.ProviderAccount{
		{ID: "p1-a1", Provider: "provider_1"},
		{ID: "p1-a2", Provider: "provider_1"},
		{ID: "p1-a3", Provider: "provider_1"},
		{ID: "p1-a4", Provider: "provider_1"},
		{ID: "p1-a5", Provider: "provider_1"},
		{ID: "p2-a1", Provider: "provider_2"},
	}}
	request := httptest.NewRequest(http.MethodPost, "/v1/chat/completions", nil)
	cfg := endpointAdapter{
		Endpoint:             "/v1/chat/completions",
		NoAccountsStatusCode: http.StatusTooManyRequests,
		Build: func(parsed parsedEndpointRequest, model string, stream bool, sessionID string) map[string]any {
			return map[string]any{"model": model, "stream": stream}
		},
	}

	account, resp, _, _, failures, _, routeErr := executeAccountRotation(
		runner,
		context.Background(),
		request,
		cfg,
		parsedEndpointRequest{Stream: false},
		auth.Result{UserID: "user_1"},
		auth.ModelValidationResult{Valid: true, Model: "unit-model"},
		nil,
		time.Now().UnixMilli(),
	)

	if routeErr != nil {
		t.Fatalf("executeAccountRotation routeErr = %+v", routeErr)
	}
	if account == nil || account.ID != "p2-a1" {
		t.Fatalf("selected account = %#v, want p2-a1", account)
	}
	if resp == nil || resp.StatusCode != http.StatusOK {
		t.Fatalf("response = %#v, want 200", resp)
	}
	if len(failures) != 5 {
		t.Fatalf("recoverable failures = %d, want 5", len(failures))
	}
	if len(runner.requested) != 6 {
		t.Fatalf("provider requests = %#v, want 6 attempts", runner.requested)
	}
}

func TestSessionIDHeaderPrecedence(t *testing.T) {
	request := httptest.NewRequest("POST", "/v1/chat/completions", nil)
	request.Header.Set("session_id", "primary")
	request.Header.Set("x-session-id", "fallback")

	if got := sessionID(request); got != "primary" {
		t.Fatalf("sessionID = %q, want primary", got)
	}
}

func TestSessionIDFallsBackToXSessionID(t *testing.T) {
	request := httptest.NewRequest("POST", "/v1/chat/completions", nil)
	request.Header.Set("x-session-id", "fallback")

	if got := sessionID(request); got != "fallback" {
		t.Fatalf("sessionID = %q, want fallback", got)
	}
}

func TestAddOpencodeRequestMetadataUsesClientIPHeaders(t *testing.T) {
	request := httptest.NewRequest(http.MethodPost, "/v1/chat/completions", nil)
	request.Header.Set("X-Forwarded-For", "203.0.113.10, 10.0.0.1")
	payload := map[string]any{}

	addOpencodeRequestMetadata(appdb.ProviderAccount{Provider: "opencode"}, payload, request)

	if payload["_realIP"] != "203.0.113.10" {
		t.Fatalf("_realIP = %#v, want 203.0.113.10", payload["_realIP"])
	}
	if payload["_projectId"] != "global" {
		t.Fatalf("_projectId = %#v, want global", payload["_projectId"])
	}
}

func TestAddOpencodeRequestMetadataUsesProjectHeader(t *testing.T) {
	request := httptest.NewRequest(http.MethodPost, "/v1/chat/completions", nil)
	request.Header.Set("X-Opencode-Project", "project_1")
	payload := map[string]any{}

	addOpencodeRequestMetadata(appdb.ProviderAccount{Provider: "opencode"}, payload, request)

	if payload["_projectId"] != "project_1" {
		t.Fatalf("_projectId = %#v, want project_1", payload["_projectId"])
	}
}

func TestAddOpencodeRequestMetadataIgnoresOtherProviders(t *testing.T) {
	request := httptest.NewRequest(http.MethodPost, "/v1/chat/completions", nil)
	request.Header.Set("X-Real-IP", "203.0.113.10")
	payload := map[string]any{}

	addOpencodeRequestMetadata(appdb.ProviderAccount{Provider: "openrouter"}, payload, request)

	if len(payload) != 0 {
		t.Fatalf("unexpected metadata for non-opencode provider: %#v", payload)
	}
}

func TestFailedCooldownUntilUsesConfiguredCooldown(t *testing.T) {
	failedAt := time.Date(2026, 5, 11, 12, 0, 0, 0, time.UTC)
	want := failedAt.Add(10 * time.Minute)

	if got := failedCooldownUntil(failedAt); !got.Equal(want) {
		t.Fatalf("failedCooldownUntil() = %v, want %v", got, want)
	}
}

func TestExecuteAccountRotationMarksCodexUsageLimitDisabledUntil(t *testing.T) {
	now := time.Now()
	resetAt := now.Add(2 * time.Hour).Unix()
	runner := &testRotationRunner{
		accounts:     []appdb.ProviderAccount{{ID: "codex-account", Provider: "codex"}},
		responseBody: `{"error":{"type":"usage_limit_reached","message":"The usage limit has been reached","plan_type":"free","resets_at":` + strconv.FormatInt(resetAt, 10) + `,"resets_in_seconds":7200}}`,
	}
	request := httptest.NewRequest(http.MethodPost, "/v1/chat/completions", nil)
	cfg := endpointAdapter{
		Endpoint:             "/v1/chat/completions",
		NoAccountsStatusCode: http.StatusTooManyRequests,
		Build: func(parsed parsedEndpointRequest, model string, stream bool, sessionID string) map[string]any {
			return map[string]any{"model": model, "stream": stream}
		},
	}

	_, _, _, _, _, _, routeErr := executeAccountRotation(
		runner,
		context.Background(),
		request,
		cfg,
		parsedEndpointRequest{Stream: false},
		auth.Result{UserID: "user_1"},
		auth.ModelValidationResult{Valid: true, Model: "gpt-5.5", Provider: strPtr("codex")},
		nil,
		now.UnixMilli(),
	)

	if routeErr == nil || routeErr.Status != http.StatusTooManyRequests {
		t.Fatalf("routeErr = %+v, want 429", routeErr)
	}
	if runner.usageLimitedAccountID != "codex-account" {
		t.Fatalf("usageLimitedAccountID = %q, want codex-account", runner.usageLimitedAccountID)
	}
	if runner.usageLimitedModel != "gpt-5.5" {
		t.Fatalf("usageLimitedModel = %q, want gpt-5.5", runner.usageLimitedModel)
	}
	if got := runner.usageLimitedUntil.Unix(); got != resetAt {
		t.Fatalf("usageLimitedUntil = %d, want %d", got, resetAt)
	}
}

func TestExecuteAccountRotationDelaysAntigravityResourceExhaustedFailureOnRecovery(t *testing.T) {
	runner := &testRotationRunner{
		accounts: []appdb.ProviderAccount{
			{ID: "ag-a1", Provider: "antigravity"},
			{ID: "p2-a1", Provider: "provider_2"},
		},
		responseBody: antigravityResourceExhaustedBody,
	}
	request := httptest.NewRequest(http.MethodPost, "/v1/chat/completions", nil)
	cfg := endpointAdapter{
		Endpoint:             "/v1/chat/completions",
		NoAccountsStatusCode: http.StatusTooManyRequests,
		Build: func(parsed parsedEndpointRequest, model string, stream bool, sessionID string) map[string]any {
			return map[string]any{"model": model, "stream": stream}
		},
	}

	account, resp, _, _, _, _, routeErr := executeAccountRotation(
		runner,
		context.Background(),
		request,
		cfg,
		parsedEndpointRequest{Stream: false},
		auth.Result{UserID: "user_1"},
		auth.ModelValidationResult{Valid: true, Model: "gemini-3-flash-preview", Provider: strPtr("antigravity")},
		nil,
		time.Now().UnixMilli(),
	)

	if routeErr != nil {
		t.Fatalf("executeAccountRotation routeErr = %+v", routeErr)
	}
	if account == nil || account.ID != "p2-a1" {
		t.Fatalf("selected account = %#v, want p2-a1", account)
	}
	if resp == nil || resp.StatusCode != http.StatusOK {
		t.Fatalf("response = %#v, want 200", resp)
	}
	if len(runner.failedAccountIDs) != 0 {
		t.Fatalf("failed accounts = %#v, want none", runner.failedAccountIDs)
	}
}

func TestExecuteAccountRotationRecordsLastAntigravityResourceExhaustedFailure(t *testing.T) {
	runner := &testRotationRunner{
		accounts: []appdb.ProviderAccount{
			{ID: "ag-a1", Provider: "antigravity"},
			{ID: "ag-a2", Provider: "antigravity"},
		},
		responseBody: antigravityResourceExhaustedBody,
	}
	request := httptest.NewRequest(http.MethodPost, "/v1/chat/completions", nil)
	cfg := endpointAdapter{
		Endpoint:             "/v1/chat/completions",
		NoAccountsStatusCode: http.StatusTooManyRequests,
		Build: func(parsed parsedEndpointRequest, model string, stream bool, sessionID string) map[string]any {
			return map[string]any{"model": model, "stream": stream}
		},
	}

	_, _, _, _, _, _, routeErr := executeAccountRotation(
		runner,
		context.Background(),
		request,
		cfg,
		parsedEndpointRequest{Stream: false},
		auth.Result{UserID: "user_1"},
		auth.ModelValidationResult{Valid: true, Model: "gemini-3-flash-preview", Provider: strPtr("antigravity")},
		nil,
		time.Now().UnixMilli(),
	)

	if routeErr == nil || routeErr.Status != http.StatusTooManyRequests {
		t.Fatalf("routeErr = %+v, want 429", routeErr)
	}
	if len(runner.failedAccountIDs) != 1 || runner.failedAccountIDs[0] != "ag-a2" {
		t.Fatalf("failed accounts = %#v, want only ag-a2", runner.failedAccountIDs)
	}
}

func TestExecuteAccountRotationUsesSharedAccountAfterOwnAccountsFail(t *testing.T) {
	runner := &testRotationRunner{
		accounts:       []appdb.ProviderAccount{{ID: "own-a1", UserID: "user_1", Provider: "provider_1"}},
		sharedAccounts: []appdb.ProviderAccount{{ID: "shared-a1", UserID: "user_2", Provider: "provider_2"}},
	}
	request := httptest.NewRequest(http.MethodPost, "/v1/chat/completions", nil)
	cfg := endpointAdapter{
		Endpoint:             "/v1/chat/completions",
		NoAccountsStatusCode: http.StatusTooManyRequests,
		Build: func(parsed parsedEndpointRequest, model string, stream bool, sessionID string) map[string]any {
			return map[string]any{"model": model, "stream": stream}
		},
	}

	account, resp, _, _, _, roaming, routeErr := executeAccountRotation(
		runner,
		context.Background(),
		request,
		cfg,
		parsedEndpointRequest{Stream: false},
		auth.Result{UserID: "user_1", RoamingEnabled: true},
		auth.ModelValidationResult{Valid: true, Model: "unit-model"},
		nil,
		time.Now().UnixMilli(),
	)

	if routeErr != nil {
		t.Fatalf("executeAccountRotation routeErr = %+v", routeErr)
	}
	if account == nil || account.ID != "shared-a1" {
		t.Fatalf("selected account = %#v, want shared-a1", account)
	}
	if resp == nil || resp.StatusCode != http.StatusOK {
		t.Fatalf("response = %#v, want 200", resp)
	}
	if roaming == nil || roaming.DebitID != "debit_1" {
		t.Fatalf("roaming reservation = %#v, want debit_1", roaming)
	}
	if got := strings.Join(runner.requested, ","); got != "own-a1,shared-a1" {
		t.Fatalf("requested = %q, want own-a1,shared-a1", got)
	}
	if len(runner.reserved) != 1 || runner.reserved[0].UserID != "user_1" {
		t.Fatalf("reserved = %#v, want one reservation for user_1", runner.reserved)
	}
	if len(runner.refunded) != 0 {
		t.Fatalf("refunded = %#v, want none", runner.refunded)
	}
}

func TestExecuteAccountRotationRefundsFailedSharedAttempt(t *testing.T) {
	runner := &testRotationRunner{
		accounts:       []appdb.ProviderAccount{{ID: "own-a1", UserID: "user_1", Provider: "provider_1"}},
		sharedAccounts: []appdb.ProviderAccount{{ID: "shared-a1", UserID: "user_2", Provider: "provider_1"}},
	}
	request := httptest.NewRequest(http.MethodPost, "/v1/chat/completions", nil)
	cfg := endpointAdapter{
		Endpoint:             "/v1/chat/completions",
		NoAccountsStatusCode: http.StatusTooManyRequests,
		Build: func(parsed parsedEndpointRequest, model string, stream bool, sessionID string) map[string]any {
			return map[string]any{"model": model, "stream": stream}
		},
	}

	_, _, _, _, _, roaming, routeErr := executeAccountRotation(
		runner,
		context.Background(),
		request,
		cfg,
		parsedEndpointRequest{Stream: false},
		auth.Result{UserID: "user_1", RoamingEnabled: true},
		auth.ModelValidationResult{Valid: true, Model: "unit-model"},
		nil,
		time.Now().UnixMilli(),
	)

	if routeErr == nil || routeErr.Status != http.StatusTooManyRequests {
		t.Fatalf("routeErr = %+v, want 429", routeErr)
	}
	if roaming != nil {
		t.Fatalf("roaming reservation = %#v, want nil on failed route", roaming)
	}
	if len(runner.reserved) != 1 || runner.reserved[0].DebitID != "debit_1" {
		t.Fatalf("reserved = %#v, want debit_1", runner.reserved)
	}
	if len(runner.refunded) != 1 || runner.refunded[0] != "debit_1" {
		t.Fatalf("refunded = %#v, want debit_1", runner.refunded)
	}
}

func TestExecuteAccountRotationStopsSharedAttemptWithoutPoints(t *testing.T) {
	runner := &testRotationRunner{
		accounts:           []appdb.ProviderAccount{{ID: "own-a1", UserID: "user_1", Provider: "provider_1"}},
		sharedAccounts:     []appdb.ProviderAccount{{ID: "shared-a1", UserID: "user_2", Provider: "provider_2"}},
		insufficientPoints: true,
	}
	request := httptest.NewRequest(http.MethodPost, "/v1/chat/completions", nil)
	cfg := endpointAdapter{
		Endpoint:             "/v1/chat/completions",
		NoAccountsStatusCode: http.StatusTooManyRequests,
		Build: func(parsed parsedEndpointRequest, model string, stream bool, sessionID string) map[string]any {
			return map[string]any{"model": model, "stream": stream}
		},
	}

	_, _, _, _, _, _, routeErr := executeAccountRotation(
		runner,
		context.Background(),
		request,
		cfg,
		parsedEndpointRequest{Stream: false},
		auth.Result{UserID: "user_1", RoamingEnabled: true},
		auth.ModelValidationResult{Valid: true, Model: "unit-model"},
		nil,
		time.Now().UnixMilli(),
	)

	if routeErr == nil || routeErr.Status != http.StatusPaymentRequired || routeErr.Code == nil || *routeErr.Code != "insufficient_points" {
		t.Fatalf("routeErr = %+v, want insufficient points", routeErr)
	}
	if got := strings.Join(runner.requested, ","); got != "own-a1" {
		t.Fatalf("requested = %q, want only own-a1", got)
	}
	if len(runner.reserved) != 0 || len(runner.refunded) != 0 {
		t.Fatalf("reserved=%#v refunded=%#v, want none", runner.reserved, runner.refunded)
	}
}

const antigravityResourceExhaustedBody = `{"error":{"code":429,"message":"Resource has been exhausted (e.g. check quota).","status":"RESOURCE_EXHAUSTED"}}`

type testRotationRunner struct {
	accounts              []appdb.ProviderAccount
	sharedAccounts        []appdb.ProviderAccount
	requested             []string
	responseBody          string
	usageLimitedAccountID string
	usageLimitedModel     string
	usageLimitedUntil     time.Time
	failedAccountIDs      []string
	reserved              []pointReservation
	refunded              []string
	insufficientPoints    bool
}

func (r *testRotationRunner) getNextAvailableAccount(_ context.Context, _ string, _ string, _ *string, exclude []string, _ auth.AccountAccess) (*appdb.ProviderAccount, bool, error) {
	excluded := map[string]struct{}{}
	for _, id := range exclude {
		excluded[id] = struct{}{}
	}
	for i := range r.accounts {
		account := &r.accounts[i]
		if _, ok := excluded[account.ID]; !ok {
			return account, true, nil
		}
	}
	return nil, false, nil
}

func (r *testRotationRunner) getNextSharedAccount(_ context.Context, _ string, _ string, _ *string, exclude []string) (*appdb.ProviderAccount, bool, error) {
	excluded := map[string]struct{}{}
	for _, id := range exclude {
		excluded[id] = struct{}{}
	}
	for i := range r.sharedAccounts {
		account := &r.sharedAccounts[i]
		if _, ok := excluded[account.ID]; !ok {
			return account, true, nil
		}
	}
	return nil, len(r.sharedAccounts) > 0, nil
}

func (r *testRotationRunner) reserveRoamingPoint(_ context.Context, userID string) (*pointReservation, bool, error) {
	if r.insufficientPoints {
		return nil, false, nil
	}
	reservation := pointReservation{UserID: userID, Amount: roamingPointCost, DebitID: "debit_" + strconv.Itoa(len(r.reserved)+1)}
	r.reserved = append(r.reserved, reservation)
	return &reservation, true, nil
}

func (r *testRotationRunner) refundRoamingPoint(_ context.Context, reservation *pointReservation) {
	if reservation == nil {
		return
	}
	r.refunded = append(r.refunded, reservation.DebitID)
}

func (r *testRotationRunner) bumpAccountRequestCount(context.Context, string, time.Time) {}

func (r *testRotationRunner) makeProviderRequest(_ context.Context, _ *http.Request, account appdb.ProviderAccount, _ map[string]any, _ bool) (*http.Response, error) {
	r.requested = append(r.requested, account.ID)
	if account.Provider == "provider_2" {
		return &http.Response{StatusCode: http.StatusOK, Body: io.NopCloser(http.NoBody)}, nil
	}
	var body io.ReadCloser = http.NoBody
	if r.responseBody != "" {
		body = io.NopCloser(strings.NewReader(r.responseBody))
	}
	return &http.Response{StatusCode: http.StatusTooManyRequests, Body: body}, nil
}

func (r *testRotationRunner) markAccountFailed(_ context.Context, accountID string, _ string, _ int, _ string) time.Time {
	r.failedAccountIDs = append(r.failedAccountIDs, accountID)
	return time.Now()
}

func (r *testRotationRunner) markAccountUsageLimited(_ context.Context, accountID, model string, disabledUntil, _ time.Time) {
	r.usageLimitedAccountID = accountID
	r.usageLimitedModel = model
	r.usageLimitedUntil = disabledUntil
}

func (r *testRotationRunner) logUsage(context.Context, usageParams) {}

func (r *testRotationRunner) isVisionModel(string) bool { return false }

func (r *testRotationRunner) isToolCallModel(string) bool { return true }

func (r *testRotationRunner) canAccountUseModel(appdb.ProviderAccount, string) bool { return true }

File: apps/proxy/test/attempts.test.ts
import { describe, expect, it } from "vitest";
import { executeAccountRotation, sessionID } from "../src/core/attempts.js";
import type { RotationRunner } from "../src/core/attempts.js";
import type { EndpointAdapter, ParsedEndpointRequest, RouteError } from "../src/core/types.js";
import type { ProviderAccount } from "../src/db/index.js";
import type { UpstreamResponse } from "../src/providers/types.js";
import { streamFromString } from "../src/providers/http.js";
import { failedCooldownUntil } from "../src/core/load-balancer.js";
import { roamingPointCost } from "../src/core/points.js";
import type { PointReservation } from "../src/core/points.js";

const antigravityResourceExhaustedBody = `{"error":{"code":429,"message":"Resource has been exhausted (e.g. check quota).","status":"RESOURCE_EXHAUSTED"}}`;

function account(id: string, provider: string, userId = ""): ProviderAccount {
  return {
    id, userId, provider, name: id, accessToken: "", refreshToken: "", expiresAt: new Date(),
    apiKey: null, projectId: null, tier: null, accountId: null, email: null, isActive: true,
    disabledUntil: null, lastUsedAt: null, status: "active",
  } as ProviderAccount;
}

class TestRotationRunner implements RotationRunner {
  accounts: ProviderAccount[];
  sharedAccounts: ProviderAccount[];
  requested: string[] = [];
  responseBody = "";
  statusByProvider: Record<string, number> = {};
  usageLimitedAccountID = "";
  usageLimitedModel = "";
  usageLimitedUntil: Date | null = null;
  failedAccountIDs: string[] = [];
  reserved: Array<{ userId: string; debitID: string }> = [];
  refunded: string[] = [];
  insufficientPoints = false;

  constructor(options: { accounts?: ProviderAccount[]; sharedAccounts?: ProviderAccount[]; responseBody?: string; statusByProvider?: Record<string, number>; insufficientPoints?: boolean } = {}) {
    this.accounts = options.accounts ?? [];
    this.sharedAccounts = options.sharedAccounts ?? [];
    if (options.responseBody !== undefined) this.responseBody = options.responseBody;
    if (options.statusByProvider !== undefined) this.statusByProvider = options.statusByProvider;
    if (options.insufficientPoints !== undefined) this.insufficientPoints = options.insufficientPoints;
  }

  async getNextAvailableAccount(_userId: string, _model: string, _provider: string | null, exclude: string[], excludeProviders: string[], _accountAccess: { mode: string; accounts: string[] }, _sessionId: string): Promise<[ProviderAccount | null, boolean, Error | null]> {
    const excluded = new Set(exclude);
    const excludedProviderSet = new Set(excludeProviders);
    for (const account of this.accounts) {
      if (excluded.has(account.id)) continue;
      if (excludedProviderSet.has(account.provider)) continue;
      return [account, true, null];
    }
    return [null, false, null];
  }

  async getNextSharedAccount(_userId: string, _model: string, _provider: string | null, exclude: string[], excludeProviders: string[]): Promise<[ProviderAccount | null, boolean, Error | null]> {
    const excluded = new Set(exclude);
    const excludedProviderSet = new Set(excludeProviders);
    for (const account of this.sharedAccounts) {
      if (excluded.has(account.id)) continue;
      if (excludedProviderSet.has(account.provider)) continue;
      return [account, true, null];
    }
    return [null, this.sharedAccounts.length > 0, null];
  }

  async reserveRoamingPoint(userId: string): Promise<[PointReservation | null, boolean, Error | null]> {
    if (this.insufficientPoints) return [null, false, null];
    const reservation: PointReservation = { userId, amount: roamingPointCost, debitID: `debit_${this.reserved.length + 1}` };
    this.reserved.push(reservation);
    return [reservation, true, null];
  }

  async refundRoamingPoint(reservation: PointReservation | null): Promise<void> {
    if (!reservation) return;
    this.refunded.push(reservation.debitID);
  }

  async bumpAccountRequestCount(): Promise<void> {}

  async makeProviderRequest(acc: ProviderAccount, _payload: Record<string, unknown>, _stream: boolean, _ctx: Record<string, unknown>): Promise<UpstreamResponse> {
    this.requested.push(acc.id);
    const status = this.statusByProvider[acc.provider];
    if (status !== undefined) {
      return { status, headers: {}, body: this.responseBody ? streamFromString(this.responseBody) : null };
    }
    if (acc.provider === "provider_2") {
      return { status: 200, headers: {}, body: null };
    }
    return { status: 429, headers: {}, body: this.responseBody ? streamFromString(this.responseBody) : null };
  }

  async markAccountFailed(accountID: string, _model: string, _status: number, _message: string): Promise<Date> {
    this.failedAccountIDs.push(accountID);
    return new Date();
  }

  async markAccountUsageLimited(accountID: string, model: string, disabledUntil: Date, _failedAt: Date): Promise<void> {
    this.usageLimitedAccountID = accountID;
    this.usageLimitedModel = model;
    this.usageLimitedUntil = disabledUntil;
  }

  async logUsage(): Promise<void> {}

  isVisionModel(): boolean {
    return false;
  }

  isToolCallModel(): boolean {
    return true;
  }
}

function cfg(): EndpointAdapter {
  return {
    endpoint: "/v1/chat/completions",
    format: "openai",
    rateLimitStatusCode: 429,
    noAccountsStatusCode: 429,
    parse: () => [{ modelParam: "", stream: false, forcedAccountID: null, reasoningRequested: false, messagesForError: undefined, paramsForError: {}, routeData: {} }, null],
    build: (parsed: ParsedEndpointRequest, model: string, stream: boolean) => ({ model, stream }),
    handleStream: async () => {},
    handleNonStream: async () => {},
  };
}

function rotate(runner: TestRotationRunner, options: { roaming?: boolean; provider?: string | null; model?: string; forced?: ProviderAccount | null } = {}) {
  return executeAccountRotation(
    runner,
    cfg(),
    { modelParam: "", stream: false, forcedAccountID: null, reasoningRequested: false, messagesForError: undefined, paramsForError: {}, routeData: {} },
    { valid: true, userId: "user_1", apiKeyId: "", modelAccessMode: "all", modelAccessList: [], accountAccessMode: "all", accountAccessList: [], roamingEnabled: options.roaming ?? false, rateLimitRules: [], error: "" },
    { valid: true, provider: options.provider ?? null, model: options.model ?? "unit-model", error: "", param: "", code: "" },
    options.forced ?? null,
    Date.now(),
    "",
  );
}

describe("attempts (ported from attempts_test.go)", () => {
  it("continues past five failures to a working account", async () => {
    const runner = new TestRotationRunner({
      accounts: [
        account("p1-a1", "provider_1"), account("p1-a2", "provider_1"), account("p1-a3", "provider_1"),
        account("p1-a4", "provider_1"), account("p1-a5", "provider_1"), account("p2-a1", "provider_2"),
      ],
    });
    const result = await rotate(runner);
    expect(result.error).toBeNull();
    expect(result.account?.id).toBe("p2-a1");
    expect(result.response?.status).toBe(200);
    expect(result.rotationFailures).toHaveLength(5);
    expect(runner.requested).toHaveLength(6);
  });

  it("sessionID header precedence", () => {
    const request = new Request("http://localhost/v1/chat/completions", { headers: { "session_id": "primary", "x-session-id": "fallback" } });
    expect(sessionID(request)).toBe("primary");
  });

  it("sessionID falls back to x-session-id", () => {
    const request = new Request("http://localhost/v1/chat/completions", { headers: { "x-session-id": "fallback" } });
    expect(sessionID(request)).toBe("fallback");
  });

  it("failedCooldownUntil uses 10 minute cooldown", () => {
    const failedAt = new Date("2026-05-11T12:00:00Z");
    expect(failedCooldownUntil(failedAt).getTime()).toBe(failedAt.getTime() + 10 * 60 * 1000);
  });

  it("marks codex usage limit disabledUntil", async () => {
    const now = new Date();
    const resetAt = Math.floor(now.getTime() / 1000) + 2 * 3600;
    const runner = new TestRotationRunner({
      accounts: [account("codex-account", "codex")],
      responseBody: `{"error":{"type":"usage_limit_reached","message":"The usage limit has been reached","plan_type":"free","resets_at":${resetAt},"resets_in_seconds":7200}}`,
    });
    const result = await rotate(runner, { provider: "codex", model: "gpt-5.5" });
    expect(result.error?.status).toBe(429);
    expect(runner.usageLimitedAccountID).toBe("codex-account");
    expect(runner.usageLimitedModel).toBe("gpt-5.5");
    expect(runner.usageLimitedUntil ? Math.floor(runner.usageLimitedUntil.getTime() / 1000) : null).toBe(resetAt);
  });

  it("delays antigravity resource exhausted failure on recovery", async () => {
    const runner = new TestRotationRunner({
      accounts: [account("ag-a1", "antigravity"), account("p2-a1", "provider_2")],
      responseBody: antigravityResourceExhaustedBody,
    });
    const result = await rotate(runner, { provider: "antigravity", model: "gemini-3-flash-preview" });
    expect(result.error).toBeNull();
    expect(result.account?.id).toBe("p2-a1");
    expect(result.response?.status).toBe(200);
    expect(runner.failedAccountIDs).toHaveLength(0);
  });

  it("records last antigravity resource exhausted failure", async () => {
    const runner = new TestRotationRunner({
      accounts: [account("ag-a1", "antigravity"), account("ag-a2", "antigravity")],
      responseBody: antigravityResourceExhaustedBody,
    });
    const result = await rotate(runner, { provider: "antigravity", model: "gemini-3-flash-preview" });
    expect(result.error?.status).toBe(429);
    expect(runner.failedAccountIDs).toEqual(["ag-a2"]);
  });

  it("uses shared account after own accounts fail", async () => {
    const runner = new TestRotationRunner({
      accounts: [account("own-a1", "provider_1", "user_1")],
      sharedAccounts: [account("shared-a1", "provider_2", "user_2")],
    });
    const result = await rotate(runner, { roaming: true });
    expect(result.error).toBeNull();
    expect(result.account?.id).toBe("shared-a1");
    expect(result.response?.status).toBe(200);
    expect(result.roaming?.debitID).toBe("debit_1");
    expect(runner.requested.join(",")).toBe("own-a1,shared-a1");
    expect(runner.reserved).toHaveLength(1);
    expect(runner.reserved[0]!.userId).toBe("user_1");
    expect(runner.refunded).toHaveLength(0);
  });

  it("falls back across providers on bad request", async () => {
    const runner = new TestRotationRunner({
      accounts: [account("p1-a1", "provider_1"), account("p1-a2", "provider_1"), account("p2-a1", "provider_2")],
      statusByProvider: { provider_1: 400 },
    });
    const result = await rotate(runner);
    expect(result.error).toBeNull();
    expect(result.account?.id).toBe("p2-a1");
    expect(result.response?.status).toBe(200);
    expect(runner.requested.join(",")).toBe("p1-a1,p2-a1");
    expect(runner.failedAccountIDs).toHaveLength(0);
    expect(result.rotationFailures).toHaveLength(0);
  });

  it("returns bad request after provider fallbacks exhausted", async () => {
    const runner = new TestRotationRunner({
      accounts: [account("p1-a1", "provider_1"), account("p1-a2", "provider_1"), account("p2-a1", "provider_2")],
      statusByProvider: { provider_1: 400, provider_2: 400 },
    });
    const result = await rotate(runner);
    expect(result.error?.status).toBe(400);
    expect(runner.requested.join(",")).toBe("p1-a1,p2-a1");
    expect(runner.failedAccountIDs).toHaveLength(0);
    expect(result.rotationFailures).toHaveLength(0);
  });

  it("does not fall back on bad request with forced provider", async () => {
    const runner = new TestRotationRunner({
      accounts: [account("p1-a1", "provider_1"), account("p2-a1", "provider_2")],
      statusByProvider: { provider_1: 400 },
    });
    const result = await rotate(runner, { provider: "provider_1" });
    expect(result.error?.status).toBe(400);
    expect(runner.requested.join(",")).toBe("p1-a1");
    expect(runner.failedAccountIDs).toHaveLength(0);
  });

  it("records last attempted account on error", async () => {
    const runner = new TestRotationRunner({
      accounts: [account("p1-a1", "provider_1"), account("p1-a2", "provider_1"), account("p2-a1", "provider_2")],
      statusByProvider: { provider_1: 429, provider_2: 429 },
    });
    const result = await rotate(runner);
    expect(result.error?.status).toBe(429);
    expect(result.error?.accountID).toBe("p2-a1");
  });

  it("omits account id when no account attempted", async () => {
    const runner = new TestRotationRunner({ accounts: [] });
    const result = await rotate(runner);
    expect(result.error).not.toBeNull();
    expect(result.error?.accountID).toBe("");
  });

  it("refunds failed shared attempt", async () => {
    const runner = new TestRotationRunner({
      accounts: [account("own-a1", "provider_1", "user_1")],
      sharedAccounts: [account("shared-a1", "provider_1", "user_2")],
    });
    const result = await rotate(runner, { roaming: true });
    expect(result.error?.status).toBe(429);
    expect(result.roaming).toBeNull();
    expect(runner.reserved).toHaveLength(1);
    expect(runner.reserved[0]!.debitID).toBe("debit_1");
    expect(runner.refunded).toEqual(["debit_1"]);
  });

  it("stops shared attempt without points", async () => {
    const runner = new TestRotationRunner({
      accounts: [account("own-a1", "provider_1", "user_1")],
      sharedAccounts: [account("shared-a1", "provider_2", "user_2")],
      insufficientPoints: true,
    });
    const result = await rotate(runner, { roaming: true });
    expect(result.error?.status).toBe(402);
    expect(result.error?.code).toBe("insufficient_points");
    expect(runner.requested.join(",")).toBe("own-a1");
    expect(runner.reserved).toHaveLength(0);
    expect(runner.refunded).toHaveLength(0);
  });
});
